File: tutorials/1d_integrator_f90/1d_integrator.f90
! uncomment for real mpi program
!program integrator

! uncomment for simulator
subroutine user_skeleton_main()

  !use mpi
  include "sstmac/libraries/mpi/sstmac_mpi_f.h"
  include "sstmac/software/libraries/compute/sstmac_compute_f.h"

  integer i, world, ierr, rank, size
  integer * 8 nintervals, nloop
  double precision start, end, x, fx, interval_size, area, total, time
  integer fakebuf

  nintervals = 10000
  start = 0
  end = 1
  area=0;
  total=0;

  call mpi_init(ierr)

  ! uncomment for real mpi program
  world = MPI_COMM_WORLD
  call mpi_comm_rank(MPI_COMM_WORLD,rank,ierr)
  call mpi_comm_size(MPI_COMM_WORLD,size,ierr)

  if(nintervals < 1) then
    stop 1
  end if

! uncomment for real mpi program
!  interval_size = (end - start) / nintervals;
!  do 10 i=rank, nintervals, size
!    x = (i + 0.5) * interval_size
!    area = area + fx(x) * interval_size
!10 continue

! uncomment for simulator
! pretend computation takes 0.1us per loop iteration
  nloop = nintervals/size
  if(rank < MOD(nintervals,size) ) then
    nloop = nloop + 1 
  end if
  time = nloop * 1e-7
  call sstmac_compute(time)

! uncomment for real mpi program
!  call mpi_reduce(area,total,1,MPI_DOUBLE_PRECISION,MPI_SUM,0,MPI_COMM_WORLD,ierr)

! uncomment for simulator
  call mpi_reduce(fakebuf,fakebuf,1,MPI_DOUBLE_PRECISION,MPI_SUM,0,world,ierr)

  if(rank .EQ. 0) then
    write (*,100) 'result is ', total
100 format (A,F8.6)
  end if

  call mpi_finalize(ierr);

end


!--------------------------------
!  function to integrate
!--------------------------------

double precision function fx(x)
  double precision x
  fx = x * x
  return
end

File: tutorials/1d_integrator_cxx/1d_integrator.cc
#include <stdlib.h>
#include <stdio.h>

#include <mpi.h>

#ifndef SSTMAC
int global_var_;
#else
global_int global_var_;
#endif

/*-------------------------------
  function to integrate
-------------------------------*/

double f_x(double x) {
  return x * x;
}


/*-------------------------------
   input variables
-------------------------------*/

int nintervals = 10000;
double start = 0;
double end = 1;


/*-------------------------------
   main 
-------------------------------*/

int main(int argc, char* argv[]) {
    int i, rank, size, nloop;
    double x, interval_size;
    double area=0;
    double total=-1.0;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    if(nintervals < 1)
        exit(EXIT_FAILURE);

#ifndef SSTMAC
    interval_size = (end - start) / nintervals;
    for(i=rank; i<nintervals; i+=size) {
        x = (i + 0.5) * interval_size;
        area += f_x(x) * interval_size;
    }
#else
    /* pretend computation takes 0.1us per loop iteration */
    nloop = nintervals/size + ((rank < (nintervals % size)) ? 1 : 0);
    SSTMAC_compute(nloop * 1e-7);
#endif
    
    MPI_Reduce(&area, &total, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

    if(rank==0) printf("result is %lf\n",total);

    MPI_Finalize();

    return EXIT_SUCCESS;
}

